src/sessions.ts

```diff
@@ -764,20 +739,37 @@ function endTransaction(
       }
     }
 
-    callback(error);
+    if (error != null) {
+      throw error;
+    }
   }
 
   if (session.transaction.recoveryToken) {
     command.recoveryToken = session.transaction.recoveryToken;
   }
 
-  const handleFirstCommandAttempt = (error?: Error) => {
+  const partiallyHandleFirstCommandAttempt = async () => {
     if (command.abortTransaction) {
       // always unpin on abort regardless of command outcome
       session.unpin();
     }
+  };
 
-    if (error instanceof MongoError && isRetryableWriteError(error)) {
+  try {
+    // send the command
+    await executeOperation(
+      session.client,
+      new RunAdminCommandOperation(command, {
+        session,
+        readPreference: ReadPreference.primary,
+        bypassPinningCheck: true
+      })
+    );
+    await partiallyHandleFirstCommandAttempt();
```

Comment:
Can we inline the logic here rather than keeping the callback?

Reply:
It would mean a small bit of code duplication, but done!